src/services/Batcher.ts

```diff
@@ -0,0 +1,144 @@
+import { MaybePromise } from '@/types'
+
+export type BatchCallbackLookup<V, R> = (value: V) => MaybePromise<R>
+export type BatchCallback<V, R> = (values: V[]) => MaybePromise<Map<V, R>> | BatchCallbackLookup<V, R>
+export type BatchOptions = {
+  maxBatchSize?: number,
+  maxWait?: number,
+}
+
+type Timer = ReturnType<typeof setTimeout>
+type Resolve<V> = (value: V) => void
+type Reject = (reason?: unknown) => void
+type BatchQueueValue<V, R> = {
+  response: Promise<R>,
+  resolve: Resolve<R>,
+  reject: Reject,
+}
+
+type BatchQueue<V, R> = Map<V, BatchQueueValue<V, R>>
+
+export class Batcher<V, R> {
+  private readonly callback: BatchCallback<V, R>
+  private readonly options: BatchOptions
+  private readonly queue: BatchQueue<V, R> = new Map()
+  private timeout: Timer | null = null
+  private waitingSince: number | null = null
+
+  public constructor(callback: BatchCallback<V, R>, options: BatchOptions = {}) {
+    this.callback = callback
+    this.options = options
+  }
+
+  public batch(value: V): Promise<R> {
+    if (this.queue.has(value)) {
+      const { response } = this.queue.get(value)!
+
+      return response
+    }
+
+    let resolve!: Resolve<R>
+    let reject!: Reject
+
+    const response = new Promise<R>((...args) => {
+      [resolve, reject] = args
+    })
+
+    this.queue.set(value, {
+      response,
+      resolve,
+      reject,
+    })
+
+    this.requestProcessQueue()
+
+    return response
+  }
+
+  private requestProcessQueue(): void {
+    if (this.shouldProcessNow()) {
+      this.processQueue()
+      return
+    }
+
+    this.waitToProcessQueue()
+  }
+
+  private waitToProcessQueue(): void {
+    if (this.waitingSince === null) {
+      this.waitingSince = Date.now()
+    }
```

Comment:
So if `waitingSince` is not null that means a queue has started but has not been batched yet. But each time this method is called it sets the timeout again, bumping/delaying the processing of the queue. But each time we "wait" we need to 
- clear the timeout (because we're about to create a new timer
- set `waitingSince` IF this is the first time we've waited